### PR TITLE
[AAP-80936] test: add non-admin user E2E tests across entire portal

### PR DESCRIPTION
**Pull request by Pavanjangle**
## Summary

Adds 20 Playwright E2E tests covering non-admin user scenarios across the entire portal. This validates the fix from PR #420 (AAP-80929) and ensures non-admin users with Org Member + JT Execute roles can access all portal pages without hitting "Insufficient privileges" errors.

### Test specs (20 tests across 3 files)

**`non-admin-permissions.spec.ts`** (6 tests) — Permission boundary checks:
- Non-admin user is authenticated and can access portal
- Does not see "Sync now" on Templates page
- Does not see "Add Template" button (Templates + EE pages)
- Does not see Administration/RBAC page in sidebar
- Cannot navigate to RBAC page directly

**`non-admin-templates.spec.ts`** (8 tests) — Templates page:
- Templates page loads without "Insufficient privileges" error
- Can see template cards or empty state
- Can use search bar and category filters
- Sees only permitted templates or empty state
- Can open template create form
- "Sync now" and "Add Template" not visible

**`non-admin-pages.spec.ts`** (6 tests) — Other portal pages:
- Can access EE catalog and Create tab
- Can access Collections page
- Can access Git Repositories page
- Can access task history page
- Sees only their own task history

### Infrastructure

- **`auth-context-user.ts`** — Worker-scoped Playwright fixture for non-admin OAuth login with retry logic
- **`aap-user-setup.ts`** — AAP API helpers to create non-admin user, assign org membership, and grant JT execute roles in global setup
- **`global-teardown.ts`** — Cleans up test user after CI runs (skipped locally to avoid deleting dev user)
- **`auth.ts`** — Extended `loginAAP()` to accept custom credentials and added fallback login verification for non-admin users
- **`.env.example`** — Documented `AAP_NONADMIN_USER_ID`, `AAP_NONADMIN_USER_PASS`, `AAP_TOKEN`, and `AAP_ORG_NAME` env vars

## RBAC note

Non-admin tests require RBAC CSV entries with action `use` (not `read`) for ansible.* permissions, because `toPermissionAction({})` defaults to `"use"` for BasicPermissions. The Jenkins CI environment needs these entries added to `default-rbac-policy.csv`:

```
p, role:default/portal-users, ansible.templates.view, use, allow
p, role:default/portal-users, ansible.execution-environments.view, use, allow
p, role:default/portal-users, ansible.collections.view, use, allow
p, role:default/portal-users, ansible.git-repositories.view, use, allow
p, role:default/portal-users, ansible.history.view, use, allow
g, user:default/e2e-nonadmin, role:default/portal-users
```

## Related issues

- https://redhat.atlassian.net/browse/AAP-80936

## Test plan

- [x] All 20 tests pass locally against dev portal (2.3 min runtime)
- [x] Verify tests pass in Jenkins CI with RBAC CSV changes applied
- [x] Validate non-admin user creation/teardown works in CI environment

## Summary by CodeRabbit

* **New Features**
  * Added end-to-end coverage for non-admin self-service pages (execution environments, collections, git repositories, history) and templates.
  * Added non-admin permission boundary validation for self-service and RBAC routes.
  * Enabled automated non-admin user provisioning/cleanup for CI runs.
* **Bug Fixes**
  * Improved Playwright login robustness and post-login authenticated UI detection with clearer failure handling.
* **Chores**
  * Updated non-admin E2E environment documentation and made Playwright global timeouts configurable; conditional global auth setup/teardown now runs only when non-admin credentials are provided.

### CODE REVIEW

**Review by audgirka — COMMENTED** [review]
Code review: 8 findings across Security, Functionality, and Quality lenses. Overall 4.2/10 — NEEDS_CHANGES. 3 Critical, 5 Major findings.

File: e2e-tests/playwright/utils/aap-user-setup.ts [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+  if (!AAP_API) return;
+  const parsed = new URL(AAP_API);
+  if (parsed.protocol !== 'https:') return;
+  if (process.env.NODE_EXTRA_CA_CERTS) return;
+
+  try {
+    const port = parsed.port || '443';
+    const cert = execSync(
+      `echo | openssl s_client -connect ${parsed.hostname}:${port} -servername ${parsed.hostname} 2>/dev/null | openssl x509`,
+      { encoding: 'utf-8', timeout: 10000 },
```
Comment (audgirka) [inline e2e-tests/playwright/utils/aap-user-setup.ts]: **Critical — Command injection**: `parsed.hostname` interpolated directly into shell command via `execSync`. If `AAP_URL` contains shell metacharacters, arbitrary commands execute.

Fix: Validate hostname with `/^[a-zA-Z0-9.-]+$/` before use, or switch to `execFileSync` with separate arguments.

File: e2e-tests/playwright/global-setup.ts [hunk 2]
```diff
@@ -1,11 +1,23 @@
...
  * This runs before all tests and saves the session to playwright/.auth/user.json
  */
 async function globalSetup(config: FullConfig) {
+  // Create non-admin test user in AAP if credentials are configured
+  if (process.env.AAP_NONADMIN_USER_ID && process.env.AAP_TOKEN) {
+    try {
+      await createNonAdminTestUser();
+    } catch {
+      console.log(
+        '[Global Setup] Non-admin user creation failed. Tests using non-admin fixture will fail.',
```
Comment (audgirka) [inline e2e-tests/playwright/global-setup.ts]: **Critical — Silent failure**: Catch block swallows the actual error message. When `createNonAdminTestUser()` fails, tests fail later with cryptic auth errors and no indication of root cause.

Fix: `catch (error) { console.log('[Global Setup] Non-admin user creation failed:', (error as Error).message); }`

File: e2e-tests/playwright/utils/aap-user-setup.ts [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+  throw new Error(`Failed to create user: HTTP ${status}`);
+}
+
+async function ensureOrgMembership(userId: number): Promise<void> {
+  const configuredOrg = process.env.AAP_ORG_NAME;
+  const query = configuredOrg
+    ? `api/gateway/v1/organizations/?name=${configuredOrg}`
+    : 'api/gateway/v1/organizations/?page_size=1';
+
```
Comment (audgirka) [inline e2e-tests/playwright/utils/aap-user-setup.ts:171]: **Critical — Non-deterministic org**: When `AAP_ORG_NAME` is unset, `page_size=1` returns whichever org the API orders first. User could be added to the wrong org, making permission tests flaky.

Fix: Either require `AAP_ORG_NAME` or skip org membership when unset.

File: e2e-tests/playwright/utils/aap-user-setup.ts [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+      '[AAP Setup] AAP_URL or AAP_TOKEN not set, skipping user creation',
+    );
+    return { id: 0, username };
+  }
+
+  console.log('[AAP Setup] Creating non-admin test user...');
+
+  // Check if user already exists
+  const existingUsers = await aapGet(
+    `api/gateway/v1/users/?username=${username}`,
```
Comment (audgirka) [inline e2e-tests/playwright/utils/aap-user-setup.ts]: **Major — URL parameter injection**: Username interpolated into query string without `encodeURIComponent`. Same pattern at lines 130, 146, 220. Special characters in the username would break the query or inject additional parameters.

Fix: Use `encodeURIComponent(username)` or `URLSearchParams`.

File: e2e-tests/playwright/utils/aap-user-setup.ts [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+      `echo | openssl s_client -connect ${parsed.hostname}:${port} -servername ${parsed.hostname} 2>/dev/null | openssl x509`,
+      { encoding: 'utf-8', timeout: 10000 },
+    );
+    if (!cert.includes('BEGIN CERTIFICATE')) return;
+
+    const tmpDir = mkdtempSync(path.join(tmpdir(), 'aap-cert-'));
+    const certPath = path.join(tmpDir, 'ca.pem');
+    writeFileSync(certPath, cert);
+    process.env.NODE_EXTRA_CA_CERTS = certPath;
+    certCleanupPath = tmpDir;
```
Comment (audgirka) [inline e2e-tests/playwright/utils/aap-user-setup.ts:45]: **Major — Untrusted cert auto-trust**: Certificate fetched from the `AAP_URL` endpoint is trusted system-wide via `NODE_EXTRA_CA_CERTS` without any validation. If `AAP_URL` pointed to a malicious server, all subsequent HTTPS connections would trust the attacker's CA.

Fix: Validate cert format before trusting. Use `{ flag: 'wx', mode: 0o600 }` on writeFile to prevent symlink attacks.

File: e2e-tests/playwright/fixtures/auth-context-user.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+        while (!loginSuccess && attempts < maxAttempts) {
+          attempts++;
+          try {
+            console.log(
+              `[Non-Admin Context] Login attempt ${attempts}/${maxAttempts}...`,
+            );
+            await loginAAP(loginPage, {
+              userId: process.env.AAP_NONADMIN_USER_ID!,
+              password: process.env.AAP_NONADMIN_USER_PASS!,
```
Comment (audgirka) [inline e2e-tests/playwright/fixtures/auth-context-user.ts]: **Major — Missing validation**: Non-null assertion `!` on `process.env.AAP_NONADMIN_USER_ID` without validation. If env vars are missing, the error will be confusing ("fill undefined into input field").

Fix: Add a guard that throws a descriptive error when credentials are not configured.

File: e2e-tests/playwright/tests/self-service/non-admin-templates.spec.ts [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        testId.toLowerCase().includes('create')
+      ) {
+        await btn.click();
+        clicked = true;
+        break;
+      }
+    }
+    if (!clicked) return;
+
+    await page.waitForTimeout(1500);
```
Comment (audgirka) [inline e2e-tests/playwright/tests/self-service/non-admin-templates.spec.ts]: **Major — Flaky test**: Hardcoded `waitForTimeout(1500)` after button click instead of waiting for navigation. On slow CI, navigation may take longer — test silently returns without assertion (false pass).

Fix: Use `page.waitForURL(/\/(create|template)/, { timeout: 5000 })` instead.

File: e2e-tests/playwright/fixtures/auth-context-user.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
+import { test as base, BrowserContext } from '@playwright/test';
```
Comment (audgirka) [inline e2e-tests/playwright/fixtures/auth-context-user.ts]: **Major — DRY violation**: This file is ~95% identical to `auth-context.ts` — only the credentials differ.

Fix: Extract a factory function `createAuthContext(getCredentials)` and derive both fixtures from it. Keeps retry logic, session management, and page fixture in one place.

**Review by audgirka — COMMENTED** [review]
## Consolidated Review — 12 Findings

Merging findings from above review and and this AI-assisted review. Deduplicated where both flagged the same issue; unique findings from each are marked.

---

### Critical / High Priority

**1. Silent admin-credential fallback when env vars are missing**
`auth-context-user.ts:37-38` — `process.env.AAP_NONADMIN_USER_ID!` produces `undefined` when unset. In `auth.ts`, `credentials?.userId || process.env.AAP_USER_ID!` then falls back to admin credentials. All non-admin permission tests silently pass as admin — validating nothing.
```ts
// Fix: add guard at top of fixture setup
if (!process.env.AAP_NONADMIN_USER_ID || !process.env.AAP_NONADMIN_USER_PASS) {
  throw new Error('AAP_NONADMIN_USER_ID and AAP_NONADMIN_USER_PASS must be set for non-admin tests');
}
```

**2. Command injection via `execSync`**
`aap-user-setup.ts:27-29` — `parsed.hostname` interpolated directly into shell command. `new URL()` does not strip shell metacharacters.
```ts
// Fix: use execFileSync with argument arrays
```

**3. Non-deterministic org selection**
`aap-user-setup.ts:145-148` — When `AAP_ORG_NAME` is unset, `page_size=1` returns whichever org the API orders first. User could be added to the wrong org, making permission tests flaky.
```ts
// Fix: require AAP_ORG_NAME or skip org membership when unset
if (!configuredOrg) {
  console.log('[AAP Setup] AAP_ORG_NAME not set, skipping org membership');
  return;
}
```

**4. Silent error swallowing in global setup**
`global-setup.ts:14-18` — Catch block discards the error message. When `createNonAdminTestUser()` fails, tests fail later with cryptic auth errors.
```ts
// Fix: log the error
} catch (error) {
  console.log('[Global Setup] Non-admin user creation failed:', (error as Error).message);
}
```

---

### Major

**5. Overly broad template card selector**
`non-admin-templates.spec.ts:21-22` — `[data-testid*="-"]` matches any element with a hyphenated test ID (buttons, containers, links). `count() > 0` is always true on a Backstage page. Tests "can see template cards" trivially pass. Used in 3 tests (lines 21, 55, 68).
```ts
// Fix: remove the overly broad selector
const templateCardSelector = '.MuiCard-root, article';
```

**6. Tautological assertions mask real failures**
`non-admin-pages.spec.ts:15-17` — After `main.waitFor({visible})`, checking `main.isVisible()` is guaranteed true. `expect(hasTabs || hasContent)` therefore always passes. Same pattern at lines 72-76 (Git Repositories test).
```ts
// Fix: assert on actual content
expect(hasTabs).toBeTruthy();
```

**7. Flaky `waitForTimeout` instead of navigation wait**
`non-admin-templates.spec.ts:96` — `waitForTimeout(1500)` after button click. On slow CI, navigation may not complete in time — test silently returns without assertion.
```ts
// Fix: wait for URL change
await page.waitForURL(/\/(create|template)/, { timeout: 5000 });
```

**8. Untrusted cert auto-trust scope**
`aap-user-setup.ts:35-36` — Certificate fetched from `AAP_URL` is trusted system-wide via `NODE_EXTRA_CA_CERTS` without format validation. Consider validating cert format and using restrictive file permissions (`{ mode: 0o600 }`).

**9. URL parameter injection**
`aap-user-setup.ts:99` (also lines 130, 220) — Username interpolated into query strings without `encodeURIComponent`. Special characters would break the query.
```ts
// Fix
`api/gateway/v1/users/?username=${encodeURIComponent(username)}`
```

---

### Minor

**10. Create/delete gate asymmetry**
`global-teardown.ts:8` — User creation has no `CI` gate, but deletion requires `process.env.CI`. Local runs against shared AAP leave orphan users. If intentional for debugging, add a comment explaining why.

**11. Duplicate tests across spec files**
`non-admin-templates.spec.ts:114-124` — "Sync now" and "Add Template" visibility tests are exact duplicates of tests in `non-admin-permissions.spec.ts`. Consider removing from the templates spec.

**12. Vacuous test passes**
`non-admin-templates.spec.ts:32-49`, `non-admin-pages.spec.ts:26-36` — Several tests wrap all logic in `if (element.isVisible())` with no else. When the element is absent, the test passes with zero assertions. Consider `test.skip()` in the else branch or asserting element presence.

---

### DRY Opportunity

I've noted earlier `auth-context-user.ts` is ~95% identical to the existing admin auth fixture. A factory function `createAuthContext(getCredentials)` would keep retry logic, session management, and page fixture in one place.

---

*Consolidated from two independent reviews. Findings 1-4 are the highest priority for correctness and debuggability.*

**Comment by Pavanjangle** [comment]
@audgirka All review comments addressed and pushed. 

Also Validated with Jenkins CI too.

Build #654 - Logs - [#654.txt](https://gitlab.cee.redhat.com/-/project/50830/uploads/12fdc855f16425b8cd90b4af0798b235/_654.txt)
Build #656 - Logs - [#656.txt](https://gitlab.cee.redhat.com/-/project/50830/uploads/0dbaa603a70b25316ff6522123a853f9/_656.txt)

File: e2e-tests/playwright/utils/aap-user-setup.ts [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+      encoding: 'utf-8',
+      timeout: 10000,
+      input: connectOut,
+    });
+    if (!cert.includes('BEGIN CERTIFICATE')) return;
+
+    const tmpDir = mkdtempSync(path.join(tmpdir(), 'aap-cert-'));
+    const certPath = path.join(tmpDir, 'ca.pem');
+    writeFileSync(certPath, cert, { mode: 0o600 });
+    process.env.NODE_EXTRA_CA_CERTS = certPath;
```
Comment (audgirka) [inline e2e-tests/playwright/utils/aap-user-setup.ts:44]: **Major — TLS fallback**: When cert extraction fails, this catch block logs and continues, relying on `NODE_TLS_REJECT_UNAUTHORIZED=0` from the environment. This silently disables all TLS verification — worse than the original auto-trust issue (one cert vs all certs).

Fix: Fail fast if cert extraction fails and `NODE_EXTRA_CA_CERTS` is not already set. Don't silently fall through to disabled TLS.

File: e2e-tests/playwright/tests/self-service/non-admin-templates.spec.ts [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+      .or(page.locator('input[placeholder*="Search"]'))
+      .first();
+
+    const isVisible = await searchInput.isVisible().catch(() => false);
+    if (!isVisible) {
+      test.skip();
+      return;
+    }
+    await searchInput.fill('test');
+    await page.waitForTimeout(1000);
```
Comment (audgirka) [inline e2e-tests/playwright/tests/self-service/non-admin-templates.spec.ts]: **Major — Flaky test**: Hardcoded `waitForTimeout(1000)` is a flaky anti-pattern. Same pattern at `non-admin-pages.spec.ts:32` and `non-admin-permissions.spec.ts:48,76`.

Fix: Use `waitForLoadState('networkidle')` or a selector-based wait like `await page.locator('...').waitFor()`.

**Review by audgirka — COMMENTED** [review]
Iteration-2 review: 5 of 8 prior findings fixed (4.2 → 8.7/10). 2 Major items remain.

**Review by sshedmake — APPROVED** [review]

**Comment by Pavanjangle** [comment]
Again tested with Jenkins CI.
 Build #728 Logs - [#728.txt](https://github.com/user-attachments/files/29782631/728.txt)

**Review by audgirka — APPROVED** [review]
LGTM
